### PR TITLE
feat: register chain and prover contract with monitoring

### PR DESCRIPTION
**Pull request by maancham**
## Description

[AXE-3574]

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

## Other Notes

[AXE-3364]: https://axelarnetwork.atlassian.net/browse/AXE-3364?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[AXE-3574]: https://axelarnetwork.atlassian.net/browse/AXE-3574?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/monitoring/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+#[derive(Error, Debug, PartialEq, IntoContractError)]
+pub enum ContractError {
+    #[error(transparent)]
+    Std(#[from] StdError),
+
+    #[error(transparent)]
+    Overflow(#[from] OverflowError),
+
+    #[error(transparent)]
+    NonEmpty(#[from] nonempty::Error),
```
Comment (cgorenflo) [inline contracts/monitoring/src/error.rs:15]: non of these are used, don't add them preemptively

Comment (maancham) [reply contracts/monitoring/src/error.rs:15]:
```
#[error(transparent)]
Std(#[from] StdError),
```
is used Chris, in checking for governance address, when loading from config:
`let config = CONFIG.load(deps.storage)?;`
if I remove `Std` error, the load fails. But I have removed the two other ones you mentioned.

File: contracts/monitoring/src/contract.rs [hunk 2]
```diff
@@ -73,14 +69,78 @@ mod tests {
...
+            },
+        );
         assert!(res.is_ok());
 
         let config = CONFIG.load(deps.as_ref().storage).unwrap();
         assert_eq!(config.governance, governance);
     }
+
+    #[test]
+    fn test_add_prover_from_goverance() {
```
Comment (cgorenflo) [inline contracts/monitoring/src/contract.rs:87]: don't start with `test_`. The functions are already clearly labeled with the test attribute. Instead add the expected outcome

File: contracts/monitoring/src/execute.rs [hunk 3]
```diff
@@ -13,17 +13,16 @@ pub fn check_governance(deps: &DepsMut, info: MessageInfo) -> Result<(), Contrac
     Ok(())
 }
 
-pub fn register_chain_contracts(
+pub fn add_chain_prover(
```
Comment (cgorenflo) [inline contracts/monitoring/src/execute.rs:16]:
```suggestion
pub fn register_prover(
```

File: contracts/monitoring/src/msg.rs [hunk 4]
```diff
@@ -10,12 +10,10 @@ pub struct InstantiateMsg {
 
 #[cw_serde]
 pub enum ExecuteMsg {
-    // Can only be called by governance account
-    RegisterChainContracts {
+    // Can only be called by governance
+    AddChainProver {
```
Comment (cgorenflo) [inline contracts/monitoring/src/msg.rs:14]:
```suggestion
    RegisterProverContract{
```

File: contracts/monitoring/src/query.rs [hunk 5]
```diff
@@ -8,3 +10,9 @@ pub fn chains_active_worker_sets(
 ) -> Vec<(ChainName, WorkerSet)> {
     todo!()
 }
+
+pub fn get_chain_provers(deps: Deps, chain_name: ChainName) -> Result<Vec<Addr>, ContractError> {
```
Comment (cgorenflo) [inline contracts/monitoring/src/query.rs:14]:
```suggestion
pub fn provers(deps: Deps, chain_name: ChainName) -> Result<Vec<Addr>, ContractError> {
```

File: contracts/monitoring/src/query.rs [hunk 6]
```diff
@@ -8,3 +10,9 @@ pub fn chains_active_worker_sets(
...
+    PROVERS_PER_CHAIN
+        .may_load(deps.storage, chain_name.clone())?
```
Comment (cgorenflo) [inline contracts/monitoring/src/query.rs:16]: no clone

File: contracts/monitoring/src/state.rs [hunk 7]
```diff
@@ -9,5 +9,5 @@ pub struct Config {
 }
 pub const CONFIG: Item<Config> = Item::new("config");
 
-// maps chain_name -> (verifier_address, gateway_address, prover_address)
-pub const CONTRACTS_PER_CHAIN: Map<ChainName, (Addr, Addr, Addr)> = Map::new("contracts_per_chain");
+// maps chain_name -> (vector of prover_addresses)
```
Comment (cgorenflo) [inline contracts/monitoring/src/state.rs:12]:
```suggestion
// maps chain name to prover addresses
```

File: contracts/monitoring/src/state.rs [hunk 8]
```diff
@@ -9,5 +9,5 @@ pub struct Config {
...
+pub const PROVERS_PER_CHAIN: Map<ChainName, Vec<Addr>> = Map::new("provers_per_chain");
```
Comment (cgorenflo) [inline contracts/monitoring/src/state.rs:13]: alternative suggestion

```suggestion
type ProverAddresses = Vec<Addr>;
pub const PROVERS_PER_CHAIN: Map<ChainName, ProverAddresses> = Map::new("provers_per_chain");
```

File: contracts/rewards/src/contract/execute.rs
Comment (cgorenflo) [inline contracts/rewards/src/contract/execute.rs:1]: looks like you made your changes on a stale state and it reverts some of the recent changes

**Review by cgorenflo — CHANGES_REQUESTED** [review]

File: contracts/rewards/src/contract/execute.rs
Comment (maancham) [reply contracts/rewards/src/contract/execute.rs:1]: I reverted all changes made to rewards. Thanks for mentioning

File: contracts/monitoring/src/error.rs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+use axelar_wasm_std_derive::IntoContractError;
+use cosmwasm_std::StdError;
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq, IntoContractError)]
+pub enum ContractError {
+    #[error(transparent)]
+    Std(#[from] StdError),
+
+    #[error("unauthorized")]
```
Comment (cgorenflo) [inline contracts/monitoring/src/error.rs:10]: can you add more detail to this error string? See `Unauthorized` errors in other contracts

File: contracts/monitoring/src/error.rs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    Unauthorized,
+
+    #[error("no provers registered for chain")]
+    NoProversRegisteredForChain,
```
Comment (cgorenflo) [inline contracts/monitoring/src/error.rs:14]:
```suggestion
    #[error("no provers registered for chain {0}")]
    NoProversRegisteredForChain(ChainName),
```
not sure if this works this way or if it needs to be explicitly transformed into a string. Either way, adding the chain name helps making the error clearer

**Review by cgorenflo — APPROVED** [review]

**Pull request merged**